### PR TITLE
Mspsds 486 improving companies house search

### PR DESCRIPTION
**Pull request by jasiekmiko**
The suggested businesses search now recognises the values entered into the `status`, `type` and `additional information` fields and adds them to the search model. With the exception of the latter field, both results coming from our DB and Companies House respect the extra restrictions.

As documented on https://regulatorydelivery.atlassian.net/browse/MSPSDS-486, companies house API is quite limited in terms of filter / narrowing down the search. Instead, we're grabbing more data from the CH and filtering it ourselves.

The searches (on New Business page and on Add Business to Case page) are now implemented with progressive enhancement, providing a non-js workflow.

### CODE REVIEW

**Comment by jasiekmiko** [comment]
No JS difference:
image

File: app/helpers/businesses_helper.rb [hunk 1]
```diff
@@ -48,4 +38,54 @@ def business_params
...
+    filters[:nature_of_business_id] = business.nature_of_business_id if business.nature_of_business_id.present?
+    Business.full_search(ElasticsearchQuery.new(query, filters, {}))
+      .paginate(per_page: 20) # Big enough to return BUSINESS_SUGGESTION_LIMIT after filters get applied
+      .records
+      .reject { |candidate| excluded_ids.include?(candidate.id) }
+      .first(BUSINESS_SUGGESTION_LIMIT)
+  end
+
+  def search_companies_house_for_similar_businesses(business)
+    type_or_status_differ = lambda do |candidate|
```
Comment (broder) [inline app/helpers/businesses_helper.rb]: Should we just inline this lambda?

File: app/helpers/businesses_helper.rb [hunk 2]
```diff
@@ -48,4 +38,54 @@ def business_params
...
+    query = [business.company_name, business.additional_information].join(' ')
+    filters = {}
+    filters[:company_type_code] = business.company_type_code if business.company_type_code.present?
+    filters[:company_status_code] = business.company_status_code if business.company_status_code.present?
+    filters[:nature_of_business_id] = business.nature_of_business_id if business.nature_of_business_id.present?
+    Business.full_search(ElasticsearchQuery.new(query, filters, {}))
+      .paginate(per_page: 20) # Big enough to return BUSINESS_SUGGESTION_LIMIT after filters get applied
+      .records
+      .reject { |candidate| excluded_ids.include?(candidate.id) }
+      .first(BUSINESS_SUGGESTION_LIMIT)
```
[the remark this answers is not in the document]
Comment (jasiekmiko) [reply app/helpers/businesses_helper.rb]: Nope. Changed to filter it at ES level :)

File: app/helpers/businesses_helper.rb (hunk 1 above)
Comment (jasiekmiko) [reply app/helpers/businesses_helper.rb]: hmm, maybe..
I arrived at this through trying to accommodate style as per https://github.com/rubocop-hq/ruby-style-guide/blob/master/README.md#single-line-blocks
can't say i love the current way, but it seems uglier without it.. I think that's my best alternative:
```
  def search_companies_house_for_similar_businesses(business)
    results = search_companies_house(business.company_name)
    results.reject { |candidate|
      (business.company_type_code.present? && business.company_type_code != candidate[:company_type_code]) ||
        (business.company_status_code.present? && business.company_status_code != candidate[:company_status_code])
        # field matched by nature_of_business_id is not available on the search models returned by companies house
    }.first(BUSINESS_SUGGESTION_LIMIT)
  end
```

File: app/assets/javascripts/businesses_search.js [hunk 3]
```diff
@@ -1,5 +1,3 @@
 function buildCompaniesHouseQuery() { // eslint-disable-line no-unused-vars
-  return {
-    q: $('.businesses-search-form .search-term').val()
-  };
+  return $('.businesses-search-form form').serialize();
```
[the remark this answers is not in the document]
Comment (jasiekmiko) [reply app/assets/javascripts/businesses_search.js]: sounds good!

File: app/controllers/investigations/businesses_controller.rb [hunk 4]
```diff
@@ -3,27 +3,23 @@ class Investigations::BusinessesController < ApplicationController
...
-    @business = Business.new
-    @business.addresses.build
+    advanced_search(@investigation.businesses.map(&:id))
   end
 
   # GET /investigations/1/businesses/new
   def new; end
 
   # GET /investigations/1/businesses/suggested
   def suggested
```
[the remark this answers is not in the document]
Comment (jasiekmiko) [reply app/controllers/investigations/businesses_controller.rb]: `suggested` is needed as a separate route, as that's what the js search hits. But have merged `search` into `new` as there was no need for both of them.

[the remark this answers is not in the document]
Comment (broder) [reply app/controllers/investigations/businesses_controller.rb]: Should we rename `suggested` to `search` to match the other `businesses_controller`?

[the remark this answers is not in the document]
Comment (jasiekmiko) [reply app/controllers/investigations/businesses_controller.rb]: Yup, will do as part of MSPSDS-491

**Pull request merged**